broadinstitute/pearl
# AR-181 Participant email preferences

This started as just "I need a way to not spam my own email when repopulating the ourhealth portal", and kind of got out of hand.  Changes include:

1. Participant profile now has two booleans for email preference
2. EmailService and NotificationDisptacher have been refactored so that NotificationDispatcher uses the async version of EmailService.processNotification.  This is better so that we're not waiting on the sendgrid API during our response loop.  The basic flow is:
  a. NotificationDisptacher creates a Notification and sets it to "READY"
  b. NotificationDispatcher calls the async version of processNotification for the emailService
  c. in a separate thread, EmailService will send (or not send) the email, and then update the Notification status in the DB accordingly.  
3. The EnrolleePopulator toggles the email preferences for test users off before populating them, so that emails aren't generated

TO TEST:
1. restart ApiAdminApp and ApiParticipantApp
2. run `./scripts/populate_portal.sh ourhealth`. 
3. confirm that no emails are sent to your account
4. Go to `http://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants/OHSALK/notifications`, and confirm you see that email sending was skipped
![image](https://user-images.githubusercontent.com/2800795/224202323-8f0f6859-b235-4f55-8230-10cce5e9e542.png)
5. register a new participant, confirm enrollment email is still sent as usual.
